### PR TITLE
gui: tx export feature

### PR DESCRIPTION
**Pull request by pythcoiner**
This PR add a feature to export transactions:
Done:
 - [x] Subscription to run the  feature in a detached thread w/ possibility to send update about the ongoing progress.
 - [x] Let user choose the path using `rfd` crate
 - [x] Add a modal that show progress of the process
 - [x] estimate the progress 
 - [x] cancel feature

### CODE REVIEW

**Comment by pythcoiner** [comment]
output format is of this kind:
```csv
Date,Label,Value,Fee,Txid,Block
2024-12-10 05:47:17,,-4556.307081010001,0,490123da60b8756ef3057489db7835ed27ba1a3a9165d1415e390c01953d0700,4474
2024-12-10 05:47:55,,-4551.73961133,0,8d14c7998237b5c363e3645397a3fa10824bf2df8a026437380821530b34fc00,4504
2024-12-10 05:48:16,,-4550.32248578,0,60b2501f9d70574e91367485b234e95ebf654e6a2510c89692f0322897684001,4508
2024-12-10 05:44:05,,0.07570141,0,4d7b801685bf5f45f585639156f0bfd45f70b6cd94b9e3fc0bf50f2d2965cf04,4414
2024-12-10 05:47:29,,-4554.703452029999,0,f121deab920efc44f91640a9f80362fe5c6c55fc9efffb0ee5dec67f590e6a05,4484
2024-12-10 05:44:05,,0.08503501,0,afb8384ca4c0842ce404c5356e049e612a9a98ddabebfb3493c80a993734fc05,4414
2024-12-10 05:47:06,,-4557.356988029999,0,05c68b3dc2a2b7853ef51e113bc78e35416c65366d1b3f0a269c88b0082b1106,4464
2024-12-10 05:47:17,,-4555.75509545,0,7cf6f654146a60b35614723447a2d0e38e5a493e9ee11b1b5fba8f2d3a01b106,4474
2024-12-10 05:47:06,,-4557.00431544,0,857885b4478df654ba5ebd4be368c9881323003d774cd0682efffe073f935f07,4464
2024-12-10 05:48:25,,-4547.699713360001,0,f366faf769ac55153d2bfd7142307fb9e4c2a38738cd6aa5ac181c0fbdb48708,4510
2024-12-10 05:46:23,,-4562.92989973,0,e50e0cb41132a70e98b94015b20aff46dac89d00bda75980e57665d906ea9d09,4424
2024-12-10 05:47:29,,-4554.39258008,0,07c577b729bddeeaf50fd09437e2d80a9e649d035af96c385918df50cde9e909,4484
2024-12-10 05:46:34,,-4562.45004093,0,4e1de4ef0ad1c61231d1f500df82f22b1654aa36120c169c40cc71b57cdb510a,4434
2024-12-10 05:47:29,,-4555.42024903,0,c65eb97af1eb0b750b7b8b99263637e8d1a91f42bdec2a9ec55e408f6ffbe90a,4484
2024-12-10 05:46:55,,-4559.26884367,0,34e0df60f41084dd8719d6ee308a6a5112672c4ccdb6b0187f958b5f38b7420b,4454
2024-12-10 05:48:16,,-4550.2741867899995,0,a2830e973c7897ffc0d5a8c5a2b6c33547ff4af47e7bca1e993a46b0ef8bb80c,4508
2024-12-10 05:48:35,"tx_label",-4546.57805264,0,75596a4a2d9125ddc9adeb559b416596921f17ae97b861473e2a045949cd1a0d,4512
2024-12-10 05:46:34,,-4562.09782068,0,b172cbfff8d68414f3a0b4419aee189849043b398edbfd8d04493ddc2367680e,4434
2024-12-10 05:47:55,,-4551.98618741,0,1ed86b3bf946675f4d89dce67ef334b60c68423be86cadb99ac1d56cae5f770e,4504
2024-12-10 05:48:35,,-4546.8183422,0,0b66304faa6c794f0538bcee2c0077aaf2b865ce99d4072da30e63e44825a50e,4512
2024-12-10 05:47:45,,-4553.24543947,0,735a85c0a29cd579f40ebf102e50c28937ee41ff7795ea4a050234b45d6aeb0e,4494
2024-12-10 05:47:17,,-4556.67268762,0,82f19fc81364a0e81a836f49975af0668c56978ad501176a9691bae95319a40f,4474
2024-12-10 05:47:17,,-4556.0801701400005,0,20613356850d17db0a7edd22e4220e6714df6c385f2bcb5cc819eeee5defa60f,4474
2024-12-10 05:46:34,,-4562.50612539,0,458e7d9042717c23f2f03b2d1f6612916ac030d23ea974a36c1dd4ca1aef9a10,4434
2024-12-10 05:47:55,,-4552.09575805,0,8160c03c683d58345c7c5ab6a3997e26f524ee5bd5d2fb6c792989c7a6dff511,4504
2024-12-10 05:47:06,,-4557.75175757,0,159ec931e9c59e451e5332f21bad87ecc6d6672cbb3c856b09e97a028db4be12,4464
2024-12-10 05:48:06,,-4550.39534451,0,25d28a012367c160b40ad009aaf565ab99e277af911295da229da9da84684d14,4506
2024-12-10 05:48:35,,-4546.88395709,0,b436128c26517fddf0f1ded028347e3f38f0b557ebc258c60bae575f8034b214,4512
2024-12-10 05:48:35,,-4547.28617754,0,a16642d97443b52810730daebb77571b7764c31bf6d4a70448bca2a7a16fca14,4512
2024-12-10 05:46:23,,-4563.71602614,0,b2d34d4e9590551a09cb54a467cc675b0ba0a4776ed7a64dfe7d0e860708e914,4424
```

File: liana-gui/src/app/export.rs [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+                    Ok(info) => info.timestamp,
+                    Err(e) => {
+                        send_error!(sender, e.into());
+                        return;
+                    }
+                };
+                // look 2 hour forward
+                let end = ((Utc::now() + Duration::hours(2)).timestamp()) as u32;
+
+                let history = daemon.list_history_txs(start, end, u32::MAX as u64).await;
```
Comment (edouardparis) [inline liana-gui/src/app/export.rs:148]: remote backend api have limits, you will need to paginate this instead of batch fetch all of it.

Comment (pythcoiner) [reply liana-gui/src/app/export.rs:148]: what is the limit? shouldn't limits be enforced by the list_history_txs() API?

Comment (edouardparis) [reply liana-gui/src/app/export.rs:148]: You can use 20, this is the default limit.

**Comment by nondiremanuel** [comment]
> we need to decide wich label select, for now i follow this logic:
> 
>     * if there is a tx label pick it
> 
>     * else if it's a single payment use the address
> 
>     * else if it's a self send show "self send"
> 
>     * else no label

Maybe we should just stick to the transaction label (if present), in order to not over-complicate the logic and create confusion due to the fact that sometimes you can find the address label (if it was a single payment) while sometimes not (if it was a batch payment). For this kind of detail, the user could refer to the "payment export" which will be more related to this kind of information detail. What do you think?

In such case the variable would be called "Tx Label" instead of a generic "Label" name.

**Comment by pythcoiner** [comment]
> Maybe we should just stick to the transaction label (if present), in order to not over-complicate the logic and create confusion due to the fact that sometimes you can find the address label (if it was a single payment) while sometimes not (if it was a batch payment). For this kind of detail, the user could refer to the "payment export" which will be more related to this kind of information detail. What do you think?
> 
> In such case the variable would be called "Tx Label" instead of a generic "Label" name.

i'm fine w/ using only the Tx label, but i think we should keep `Label` name for this column, as it is the name already used by sparow (and electrum?)

**Comment by nondiremanuel** [comment]
> > Maybe we should just stick to the transaction label (if present), in order to not over-complicate the logic and create confusion due to the fact that sometimes you can find the address label (if it was a single payment) while sometimes not (if it was a batch payment). For this kind of detail, the user could refer to the "payment export" which will be more related to this kind of information detail. What do you think?
> > In such case the variable would be called "Tx Label" instead of a generic "Label" name.
> 
> i'm fine w/ using only the Tx label, but i think we should keep `Label` name for this column, as it is the name already used by sparow (and electrum?)

I'm ok with it.

File: liana-gui/src/export.rs [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+                        break;
+                    }
+                    if history_txs.len() == limit as usize {
+                        let first = if let Some(t) = history_txs.first().expect("checked").time {
+                            t
+                        } else {
+                            send_error!(sender, TxTimeMissing);
+                            return;
+                        };
+                        let last = if let Some(t) = history_txs.first().expect("checked").time {
```
Comment (edouardparis) [inline liana-gui/src/export.rs:215]: should it be last() here instead of first() ?

File: liana-gui/src/export.rs [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+    Running,
+    Stopped,
+}
+
+#[derive(Debug, Clone)]
+pub enum ExportProgress {
+    Started(Arc<Mutex<JoinHandle<()>>>),
+    Progress(f32),
+    Ended,
+    Finnished,
```
Comment (jp1ac4) [inline liana-gui/src/export.rs:96]: Typo: `Finished`

File: liana-gui/src/export.rs [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+                            send_error!(sender, e.into());
+                            return;
+                        }
+                    };
+                    let dl = map.len() + history_txs.len();
+                    if dl > 0 {
+                        let progress = (dl as f32) / (total_txs as f32) * 80.0;
+                        send_progress!(sender, Progress(progress));
+                    }
+                    // all txs have been fetch
```
Comment (jp1ac4) [inline liana-gui/src/export.rs:204]: nit: "fetched"

File: liana-gui/src/export.rs [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+                            return;
+                        };
+                        let last = if let Some(t) = history_txs.first().expect("checked").time {
+                            t
+                        } else {
+                            send_error!(sender, TxTimeMissing);
+                            return;
+                        };
+                        // limit too low, all tx are in the same timestamp
+                        // whe must increase limit and retry
```
Comment (jp1ac4) [inline liana-gui/src/export.rs:222]: nit: "we"

File: liana-gui/src/app/state/export.rs [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+use crate::app::{
+    message::Message,
+    view::{self, export::export_modal},
+};
+use crate::daemon::Daemon;
+use crate::export::{self, get_path, ExportProgress};
+
+#[derive(Debug, Clone)]
+pub enum ExportMessage {
```
Comment (jp1ac4) [inline liana-gui/src/app/state/export.rs:18]: Could `ExportMessage`, `ExportState` and perhaps some other parts of the code here also be moved to `liana-gui/src/export.rs`? They seem quite general and could also be reused in the installer, for example.

File: liana-gui/src/app/view/export.rs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            ExportState::Init => "".to_string(),
+            ExportState::ChoosePath => {
+                "Select the path you want to export in the popup window...".into()
+            }
+            ExportState::Path(_) => "".into(),
+            ExportState::Started => "Starting export...".into(),
+            ExportState::Progress(p) => format!("Progress: {}%", p.round()),
+            ExportState::TimedOut => "Export failed: timeout".into(),
+            ExportState::Aborted => "Export canceled".into(),
+            ExportState::Ended => "Export successfull!".into(),
```
Comment (jp1ac4) [inline liana-gui/src/app/view/export.rs:47]: typo: "successful"

**Review by jp1ac4 — COMMENTED** [review]
I'm continuing to test & review, but adding some initial comments.

File: liana-gui/src/export.rs [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+                    let txid_label = tx.labels().get(&txid).cloned();
+                    let mut label = if let Some(txid) = txid_label {
+                        txid
+                    } else {
+                        "".to_string()
+                    };
+                    if !label.is_empty() {
+                        label = format!("\"{}\"", label);
+                    }
+                    let fee = tx.fee_amount.unwrap_or(Amount::ZERO).to_sat() as i128;
```
Comment (jp1ac4) [inline liana-gui/src/export.rs:268]: Should the fee also be shown in BTC to be consistent with the Value column?

File: liana-gui/src/app/state/transactions.rs [hunk 9]
```diff
@@ -246,11 +259,27 @@ impl State for TransactionsPanel {
                     );
                 }
             }
-            _ => {
-                if let Some(modal) = &mut self.create_rbf_modal {
-                    return modal.update(daemon, _cache, message);
+            Message::View(view::Message::Export(ExportMessage::Open)) => {
+                if let TransactionsModal::None = &self.modal {
+                    self.modal = TransactionsModal::Export(ExportModal::new(daemon));
+                    if let TransactionsModal::Export(m) = &mut self.modal {
```
Comment (jp1ac4) [inline liana-gui/src/app/state/transactions.rs:265]: nit: As this `if` condition should be true, perhaps you could store `ExportModal::new(daemon)` in a variable and then use it to call `launch()`.

File: liana-gui/src/app/state/transactions.rs [hunk 10]
```diff
@@ -246,11 +259,27 @@ impl State for TransactionsPanel {
...
+                        return m.launch();
+                    }
                 }
             }
+            Message::View(view::Message::Export(ExportMessage::Close)) => {
+                // Message::Export(ExportMessage::Close) => {
```
Comment (jp1ac4) [inline liana-gui/src/app/state/transactions.rs:271]: Can this line be removed?

File: liana-gui/src/export.rs [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+        State {
+            receiver,
+            sender: Some(sender),
+            handle: None,
+            daemon,
+            path,
+        }
+    }
+
+    pub async fn start(&mut self) {
```
Comment (jp1ac4) [inline liana-gui/src/export.rs:151]: Is there a way to move the logic in this method to a module specific to the transactions export and keep only the general parts here?

File: liana-gui/src/export.rs [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+#[derive(Debug, Clone)]
+pub enum Error {
+    Io(String),
+    HandleLost,
+    UnexpectedEnd,
+    JoinError(String),
+    ChannelLost,
+    NoParentDir,
+    Daemon(String),
+    TxTimeMissing,
```
Comment (jp1ac4) [inline liana-gui/src/export.rs:93]: These final two, especially the latter, might not be relevant to all exports.

**Review by jp1ac4 — COMMENTED** [review]
I've tested the export with a couple of wallets and it's looking good. Added some comments.

File: liana-gui/src/export.rs (hunk 11 above)
Comment (pythcoiner) [reply liana-gui/src/export.rs:151]: i've try this before but it was a real pain, i'd prefer keep this way for now (some weird issue about async)

File: liana-gui/src/app/state/transactions.rs [hunk 13]
```diff
@@ -28,20 +28,30 @@ use crate::{
...
     Daemon,
 };
 
+use super::export::ExportModal;
+
+#[derive(Debug)]
+pub enum TransactionsModal {
+    CreateRbf(CreateRbfModal),
+    Export(ExportModal),
+    None,
```
[the remark this answers is not in the document]
Comment (pythcoiner) [reply liana-gui/src/app/state/transactions.rs:45]: consistency, see [receive panel](https://github.com/wizardsardine/liana/blob/ecce76ab3f260ce405659f4e043ae99b97aee836/liana-gui/src/app/state/receive.rs#L31)

File: liana-gui/src/app/state/transactions.rs (hunk 9 above)
Comment (pythcoiner) [reply liana-gui/src/app/state/transactions.rs:265]: modal has to be stored in the state, but i can `.expect()` it if you prefer

File: liana-gui/src/export.rs (hunk 12 above)
Comment (pythcoiner) [reply liana-gui/src/export.rs:93]: i think it's fine for now, this + adding generics should be added when we implement a second type export imho

File: liana-gui/src/app/state/transactions.rs (hunk 9 above)
Comment (jp1ac4) [reply liana-gui/src/app/state/transactions.rs:265]: I had in mind something like:
```diff
diff --git a/liana-gui/src/app/state/transactions.rs b/liana-gui/src/app/state/transactions.rs
index ad90d5d3..a04b3829 100644
--- a/liana-gui/src/app/state/transactions.rs
+++ b/liana-gui/src/app/state/transactions.rs
@@ -261,10 +261,10 @@ impl State for TransactionsPanel {
             }
             Message::View(view::Message::Export(ExportMessage::Open)) => {
                 if let TransactionsModal::None = &self.modal {
-                    self.modal = TransactionsModal::Export(ExportModal::new(daemon));
-                    if let TransactionsModal::Export(m) = &mut self.modal {
-                        return m.launch();
-                    }
+                    let modal = ExportModal::new(daemon);
+                    let cmd = modal.launch();
+                    self.modal = TransactionsModal::Export(modal);
+                    return cmd;
                 }
             }
             Message::View(view::Message::Export(ExportMessage::Close)) => {
```
I think `.expect()` would also be fine or also to keep it as it is now.

You can remove the `mut` now following your change to the `launch()` method.

File: liana-gui/src/export.rs (hunk 11 above)
Comment (jp1ac4) [reply liana-gui/src/export.rs:151]: OK sure, we can come back to that the next time we need to add an export.

File: liana-gui/src/export.rs (hunk 12 above)
Comment (jp1ac4) [reply liana-gui/src/export.rs:93]: Sounds good.

File: liana-gui/src/app/state/transactions.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (jp1ac4) [reply liana-gui/src/app/state/transactions.rs:45]: OK thanks.

File: liana-gui/src/app/state/transactions.rs (hunk 9 above)
Comment (pythcoiner) [reply liana-gui/src/app/state/transactions.rs]: i've dropped the `mut`

**Comment by jp1ac4** [comment]
This needs a rebase on master.

**Comment by pythcoiner** [comment]
> This needs a rebase on master.

rebased

**Review by jp1ac4 — APPROVED** [review]
Tested ACK 22f48753a367187a1d08bc34d20e5b75bdfc01b8.

**Pull request merged**